Add flag to check if digest already exists before pushing a new tag

zbarahal opened the pull request:
Amazon ECR imposes a 100 tag per image limit. Our existing monorepo build setup pushes a new tag for each service each time we rebuild the repository regardless of if that service was affected by the commit. This has caused some of our images on less frequently changed services to bump into the limit.

This change adds a flag to rules_docker to allow checking if the digest is already in the repository before pushing a new tag

zbarahal committed:
Add flag to check if digest already exists before pushing a new tag

nathanielmiller20 commented on container/go/cmd/pusher/pusher.go:
```diff
@@ -20,12 +20,13 @@ import (
 	"github.com/google/go-containerregistry/pkg/authn"
 	"github.com/google/go-containerregistry/pkg/name"
-	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1"
```
A little weird this was explicitly done before. Any chance they're following a standard we don't internally that requires this?

zbarahal replied:
reverting. this got auto done by my IDE because it's not needed but I'll leave it for consistency.

nathanielmiller20 commented on container/go/cmd/pusher/pusher.go:
```diff
@@ -124,6 +127,27 @@ func main() {
 }
 
+func digestExists(dst string ,img v1.Image) (bool, error) {
+	// Try to create a digest for the image and then check if that digest exists in the repository
```
Should this comment be on `digestExists`? Something like:

`// digestExists checks whether the image's digest already exists in the repository`

nathanielmiller20 commented on container/go/cmd/pusher/pusher.go:
```diff
@@ -124,6 +127,27 @@ func main() {
 	log.Printf("Successfully pushed %s image from %s to %s", *format, imgSrc, formattedDst)
 }
 
+func digestExists(dst string ,img v1.Image) (bool, error) {
```
Nit: swap comma with space before it

i replied:
surprised gofmt didn't pick this up. be sure to run that before submitting

zbarahal replied:
I may have broken something in my IDE goofing around with gomod. Will double check that it's gofmted

nathanielmiller20 commented on container/go/cmd/pusher/pusher.go:
```diff
@@ -46,6 +48,7 @@ const (
 	manifestFile = "manifest.json"
 	// indexManifestFile is the filename of image manifest config in OCI format.
 	indexManifestFile = "index.json"
+	ManifestUnknownError = "MANIFEST_UNKNOWN: manifest unknown"
```
Nit: this can probably be private? Don't see it used elsewhere. Also, is it a standard that every const has a comment? If so, we should add one.

nathanielmiller20 replied:
One question I just thought of: any chance these messages differ by registry provider?

zbarahal replied:
very good question. I only tested iwth a local docker registry. I will try with ECR as well

nathanielmiller20 commented on container/go/cmd/pusher/pusher.go:
```diff
@@ -135,6 +159,16 @@ func push(dst string, img v1.Image) error {
+	if *onlyPushChanged {
+		exists, err := digestExists(dst, img)
+		if err != nil {
+			log.Printf("Error checking if digest already exists %v. Still pushing", err)
```
Should we fail the build instead? I guess this usually isn't a big deal. Should we note this behavior in the option explainer? That seems too verbose. Mostly a devil's advocate comment, I think.

nathanielmiller20 commented on container/new_push.bzl:
```diff
@@ -214,7 +222,7 @@ new_container_push = rule(
             allow_single_file = True,
         ),
-    }, _layer_tools),
+            }, _layer_tools),
```
Nit: This doesn't seem right.

nathanielmiller20 reviewed (approved):
LGTM, bunch of nit comments.

zbarahal committed:
Clean up formatting

zbarahal committed:
Add comment for const

zbarahal committed:
Use more generic form of error

smukherj1 committed:
Implement join_layers.py in Go (#1111)

* Start work to implement join_layers.py in Go

Changes:
1. Refactored existing legacyImage to generate manifest on demand instead
   of always expecting a manifest to match the behavior of join_layers.py.
2. Added support for reading uncompressed layers.
3. Added a blank join_layers Go binary that only parses flags.
4. Modified layer_tools.bzl to call the Go binary instead of the python
   binary.

All container_image rules currently fail when an output image tarball is
requested because the join_layers Go binary doesn't produce an output
file.

* Implement most of join_layers.go except foreign layers and base images

Foreign layers currently create invalid image and they fail validation
in the reader. Need to reach out to go-containerregistry team for
guidance.

Not reading base image layers yet which is TODO.

* Add support for foreign layers & loading layers from base image tarballs

Foreign layer support just involved removing the image validation code
which doesn't support foreign layers and is too runtime expensive
anyway.

* Ignore manifest layers that are not foreign layers

* Fix rename image test

Go join_layers requires the registry name to be specified as well when
renaming images

* Fix stamper & loading from base image tarball

* Default join layers to py & add tests for Go implementation

* Fix stamp-info-file arg

* Address review comments

* Address Starlark review comments

* Fix

* buildifier

zbarahal committed:
Add integration tests for only_push_changed flag

zbarahal committed:
Merge branch 'master' into digest_check_push

alex1545 committed:
Enable specifying a tag for the checker image in configs_test (#1115)

* Enable specifying a tag for the checker image in configs_test

* fix

* address comment

zbarahal committed:
Run buildifier

zbarahal committed:
Merge branch 'master' into digest_check_push

zbarahal committed:
Fix dict ordering

zbarahal committed:
More buildifier changes

zbarahal committed:
Remove unnecessary mandatory field

zbarahal committed:
Rename flag to be more clear as to purpose

zbarahal committed:
Use existing error and clean up join

zbarahal committed:
Use correct coding style for comments/errors

zbarahal committed:
Add log message for not pushing

zbarahal committed:
Update bazel deps for transport package

zbarahal committed:
Fix import sort order

zbarahal committed:
Remove bin that got checked in

zbarahal committed:
Buildifier

zbarahal commented:
was a draft for a public pr. that one got merged
